### ISSUE TITLE
incorrect GPS altitude on OSD > 500 meters

### ISSUE THREAD

**Issue by rozhkovets** [iNavFlight/inav#3769]
## Current Behavior
video - https://youtu.be/BRm9ZRwkwMs
blackbox log - https://drive.google.com/drive/folders/1RSj63lUuZ7BKwiF2JgK7yOaFoTz_aVxs?usp=sharing

## Steps to Reproduce
1. inav 2.0.0 stable
2. fly and climb to the altitude ~>512

## Expected behavior
With increasing altitude the gps altitude value on osd becomes incorrect, decreases to negative values or disappears completely.
With decrease in altitude after a while everything comes to normal.

## Additional context
dump https://pastebin.com/4BM110kD

---
INAV/MATEKF405OSD 2.0.0 Aug 20 2018 / 19:20:20 (dbdd1656a)

**Comment by digitalentity** [comment]
Impossible to troubleshoot without blackbox log. People have flown RC6 over 2km altitude without issue.

**Comment by rozhkovets** [comment]
I updated the firmware to 2.0.0 with a complete chip erasing. Problem still exists. Attached a new video (2:53 fly time), blackbox log and google eath kmz.
video - https://youtu.be/BRm9ZRwkwMs
blackbox log and kmz - https://drive.google.com/drive/folders/1RSj63lUuZ7BKwiF2JgK7yOaFoTz_aVxs?usp=sharing

**Comment by digitalentity** [comment]
Looks like this is specific to `set gps_provider = NMEA`

```
201612284, 2, 12, 58.4731000, 56.4704116, 652, 16.20, 89.0, 50, 100, 100
201693242, 2, 12, 58.4731000, 56.4704116, 652, 16.20, 89.0, 50, 100, 100
201812421, 2, 12, 58.4731000, 56.4704666, 654, 15.27, 93.9, 50, 100, 100
202012568, 2, 12, 58.4730966, 56.4705166, 1, 14.50, 98.6, 50, 100, 100
202212821, 2, 12, 58.4730916, 56.4705650, 3, 13.78, 103.7, 50, 100, 100
202412958, 2, 12, 58.4730850, 56.4706083, 5, 13.16, 109.4, 50, 100, 100
202613149, 2, 12, 58.4730766, 56.4706500, 6, 12.65, 114.8, 50, 100, 100
202813313, 2, 12, 58.4730666, 56.4706883, 8, 12.13, 122.0, 50, 100, 100
```

After 654m absolute altitude GPS provider started reporting incorrect altitude. 654m = 65400cm which gives us a hint about 16-bit overflow in either GPS itself of NMEA decoder.

**Comment by digitalentity** [comment]
```
typedef struct gpsDataNmea_s {
    bool fix;
    int32_t latitude;
    int32_t longitude;
    uint8_t numSat;
    uint16_t altitude;
    uint16_t speed;
    uint16_t ground_course;
    uint16_t hdop;
    uint32_t time;
    uint32_t date;
} gpsDataNmea_t;
```

Got it. Looks like nobody flew NMEA higher than 655.35m MSL altitude 🤣

**Comment by stronnag** [comment]
Or lower than MSL ...

**Comment by digitalentity** [comment]
Yes, NMEA driver doesn't even consider negative altitudes

**Comment by digitalentity** [comment]
#3778

**Comment by fiam** [comment]
Fixed in #3778
